cmd/restic/cmd_cleanup.go

```diff
@@ -0,0 +1,372 @@
+package main
+
+import (
+	"golang.org/x/sync/errgroup"
+
+	"github.com/restic/restic/internal/crypto"
+	"github.com/restic/restic/internal/debug"
+	"github.com/restic/restic/internal/errors"
+	"github.com/restic/restic/internal/pack"
+	"github.com/restic/restic/internal/repository"
+	"github.com/restic/restic/internal/restic"
+
+	"github.com/spf13/cobra"
+)
+
+var cmdCleanup = &cobra.Command{
+	Use:   "cleanup [flags]",
+	Short: "Cleanup unused data",
+	Long: `
+The "cleanup" cleans up data in index and pack files
+that is not referenced in any snapshot file.
+
+When calling this command without flags, only packs
+that are completely unused are deleted and index files are cleaned.
+You can specify additional conditions to repack
+packs that are only partly used or too small.
+These packs will be downloaded and uploaded again which can be
+quite time-consuming for remote repositories.
+`,
+	DisableAutoGenTag: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return runCleanup(cleanupOptions, globalOptions)
+	},
+}
+
+// CleanupIndexOptions collects all options for the cleanup command.
+type CleanupOptions struct {
+	DryRun            bool
+	DataUnusedPercent int8
+	DataUnusedSpace   int64
+	DataUsedSpace     int64
+	TreeUnusedPercent int8
+	TreeUnusedSpace   int64
+	TreeUsedSpace     int64
+	RepackMixed       bool
+}
+
+var cleanupOptions CleanupOptions
+
+func init() {
+	cmdRoot.AddCommand(cmdCleanup)
+
+	f := cmdCleanup.Flags()
+	f.BoolVarP(&cleanupOptions.DryRun, "dry-run", "n", false, "do not delete anything, just print what would be done")
+	f.Int8Var(&cleanupOptions.DataUnusedPercent, "data-unused-percent", -1, "if > 0, repack data packs with >= given % unused space")
+	f.Int64Var(&cleanupOptions.DataUnusedSpace, "data-unused-space", -1, "if > 0, repack data packs with >= given bytes unused space")
+	f.Int64Var(&cleanupOptions.DataUsedSpace, "data-used-space", -1, "repack data packs with <= given bytes used space")
+	f.Int8Var(&cleanupOptions.TreeUnusedPercent, "tree-unused-percent", -1, "if > 0, repack tree packs with >= given % unused space")
+	f.Int64Var(&cleanupOptions.TreeUnusedSpace, "tree-unused-space", 1<<20, "if > 0, repack tree packs with >= given bytes unused space")
+	f.Int64Var(&cleanupOptions.TreeUsedSpace, "tree-used-space", 1<<16, "repack tree packs with <= given bytes used space")
```

Comment:
> And in essence, if the index and tree packs are cached, this would allow a prune run to complete without downloading any data from the backend.

This was exactly what I had in mind. And this is a really cool feature if you have cold storage in mind where you only want to download anything in the restore case.

I agree that using `--repack-trees-only` is sufficient and much clearer for users. I'll change this!